### PR TITLE
Increase variable explorer window size

### PR DESCRIPTION
**Pull request by jsh9**
## Description of Changes

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

Related issue: https://github.com/spyder-ide/spyder/issues/10745

The current variable explorer window sizes are small in today's higher resolution monitors. So I'm increasing their sizes.

Current size of a data frame viewer:

image

New size (just an example):

image

I'm waiting for @dalthviz 's help on implementing a dynamic window sizing, and I'll push up another commit to this PR.

### Issue(s) Resolved

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by dalthviz** [comment]
Hi @jsh9 about a dynamic window sizing maybe we could:

1. Create a new method in the dialog class. The method will get as params the info about the geometry of the screen and the proportions for width and height (it will do the resize of the dialog too). Something like this:

```python
def set_dynamic_width_and_height(self, screen_geometry, width_proportion=0.5,
    height_proportion=0.5):
    """
    Update width and height using an updated screen geometry.
    Use a proportion for the width and height of the dialog.
    """
    screen_width = screen_geometry.width()
    screen_height = screen_geometry.height()
    self.resize(screen_width * width_proportion, screen_height * height_proportion)
```
2. Get the `QScreen` instance of the window of the dialog in the constructor (or maybe in a setup method since some elements of the call chain could be `None` inside of the constructor, I'm not sure). Something like this:
```python
screen = self.window().windowHandle().screen().geometry()
```
3. Connect the new method created at `1.` to the `geometryChanged` signal of the QScreen instance that we get at `2.`. Something like this:
```python
screen.geometryChanged.connect(self.set_dynamic_width_and_height)
```
Hope the elements above give you some hints around the things that can be done for the dynamic resizing. Let me know if you have questions or need more help :)

**Comment by jsh9** [comment]
Hi @dalthviz , I add the following lines to the end of the constructor (for example, the constructor of `TextEditor(QDialog)`:
```python
    window = self.window()
    windowHandle = window.windowHandle()
    screen = windowHandle.screen()
    geometry = screen.geometry()
    screen.geometryChanged.connect(self.set_dynamic_width_and_height)
```

However, when I create a `str` variable and tried to view it, I got a null pointer exception:
```
    screen = windowHandle.screen()
AttributeError: 'NoneType' object has no attribute 'screen'
```
This indicates that `self.window().windowHandle()` is `None`.

So how can write this "setup method" that you mentioned? Thanks.

**Comment by dalthviz** [comment]
@jsh9 rethinking this, maybe a better approach could be the creation of a dialog base class. The base class should look something like this:

```python
class BaseDialog(QDialog):
    def set_dynamic_width_and_height(self, screen_geometry, width_proportion=0.5,
        height_proportion=0.5):
        """
        Update width and height using an updated screen geometry.
        Use a proportion for the width and height of the dialog.
        """
        screen_width = screen_geometry.width()
        screen_height = screen_geometry.height()
        self.resize(screen_width * width_proportion, screen_height * height_proportion)

    def show(self):
        super(BaseDialog, self).show()
        window = self.window()
        windowHandle = window.windowHandle()
        screen = windowHandle.screen()
        geometry = screen.geometry()
        self.set_dynamic_width_and_height(geometry)
        screen.geometryChanged.connect(self.set_dynamic_width_and_height)
```
And then the dialogs need to inherit from this base class. For example, something like:

```Python
class TextEditor(BaseDialog):
    ...
```

**Comment by jsh9** [comment]
Hi @dalthviz , I implemented your suggestion.  Please take a look at my changes.

I've tested the result on 3840x2160, 2560x1440, and 1366x768 resolutions, and they all have the expected behavior (i.e., the variable explorer window shows 50% of the screen height and 50% of the screen width).

There is one small remaining issue though: on higher resolution (for example 3840x2160 and 2560x1440), the initial position of the explorer window is always on the lower-right of the screen. Since we've hard-coded the relative window size to be 50%, it isn't an issue. But it would be better if we can center these windows. How do you think we can achieve this?

Thanks again!

**Comment by dalthviz** [comment]
@jsh9 to center the dialogs we could add the end of the `set_dynamic_width_and_height` method something like this:

```python
        x = screen_geometry.center().x() - self.width() / 2
        y = screen_geometry.center().y() - self.height() / 2
        self.move(x, y)
```

File: spyder/plugins/variableexplorer/widgets/arrayeditor.py [hunk 1]
```diff
@@ -40,7 +40,7 @@
                               to_text_string)
 from spyder.utils import icon_manager as ima
 from spyder.utils.qthelpers import add_actions, create_action, keybinding
-
+from .basedialog import BaseDialog
```
Comment (dalthviz) [inline spyder/plugins/variableexplorer/widgets/arrayeditor.py]: @jsh9 I think this import should be something like `spyder.plugins.variableexplorer.widgets.basedialog import BaseDialog` (and should be changed also for the other dialogs)

Comment (jsh9) [reply spyder/plugins/variableexplorer/widgets/arrayeditor.py]: Fixed. Thank you.

**Comment by jsh9** [comment]
Hi @dalthviz , I made some changes according to your suggestions. Thank you!

Now I have three questions remaining:

1. Is there a way to apply an offset to the explorer windows, if more than one is open? This is a nice-to-have, and maybe out of the scope of this PR. But if there is a simple way to solve it, maybe I can just put it in as well.
2. I'd also like to change the size of the object editor, but it inherits from `QObject`. `QOobject`'s documentation (https://doc.qt.io/qtforpython/PySide2/QtCore/QObject.html) is quite complicated  to me. Do you have a good suggestion?
3. Some of the CI checks failed for no apparent reasons. I looked at the logs of those that failed, and everything looked fine.

**Comment by dalthviz** [comment]
Hi @jsh9 about your questions:

1. I would need to think about that but, at least for now, I'm unsure of a proper way to achieve such behavior (maybe we could instead change the proportion per dialog reimplementing in each one the `set_dynamic_..` method, probably calling `super` but providing different default values for the height and width proportions for each dialog). What do you think @ccordoba12 ?

2. If you refer to the new `Object Explorer`, the dialog is at `spyder\spyder\plugins\variableexplorer\widgets\objectexplorer\objectexplorer.py`. Maybe you checked the object editor module (at `spyder\spyder\plugins\variableexplorer\widgets\objecteditor.py`) but that is used to create dialogs (is not a dialog itself).

3. Could you try to merge with master?

Thanks for all the work in this one @jsh9 👍. Also, could you post here a preview of how things are looking (you can create a `.gif` with something like [LICEcap](https://www.cockos.com/licecap/)) Thanks!

**Comment by jsh9** [comment]
Hi @dalthviz , I've made `ObjectExplorer` and `ImportWizart` to inherit from `BaseDialog`. I've also pulled in the latest upstream master branch.

As for the "automatic window offser" feature, I'd say let's put it in the future wish list?

Below is the screen shot of the new window size. I manually dragged different variable explorer windows so that we can see all of the in a single screenshot:
image

**Comment by ccordoba12** [comment]
@jsh9, this is looking pretty good. Thanks for your help with it!

We'd like to include your contribution in our next release in the Spyder 4 series. For that, please remove your last comment (i.e. your merge with master) and make your branch to derive from our `4.x` branch (instead of master), with the following command:

    git rebase --onto 4.x master variable_explorer_window

If you decline to do that, your work will be part of Spyder 5, to be released much later in 2020.

**Comment by jsh9** [comment]
Hi @ccordoba12 , I reverted my merge commit, and did what you asked.  I'm not sure I did the entirely correct thing, because I'm not a git expert at all. Could you please take a look?

If you could make changes directly, it will be easier.  Thanks!

**Comment by dalthviz** [comment]
Closing in favor of #10976

**Pull request closed without merging**